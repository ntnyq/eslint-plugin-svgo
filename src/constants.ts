/**
 * @see {@link https://svgo.dev/docs/preset-default/}
 */
export const PLUGIN_PRESET_DEFAULT = 'preset-default'

/**
 * @see {@link https://svgo.dev/docs/preset-default/}
 */
// @keep-sorted
export const PLUGINS_PRESET_DEFAULT = [
  'cleanupAttrs',
  'cleanupEnableBackground',
  'cleanupIds',
  'cleanupNumericValues',
  'collapseGroups',
  'convertColors',
  'convertEllipseToCircle',
  'convertPathData',
  'convertShapeToPath',
  'convertTransform',
  'inlineStyles',
  'mergePaths',
  'mergeStyles',
  'minifyStyles',
  'moveElemsAttrsToGroup',
  'moveGroupAttrsToElems',
  'removeComments',
  'removeDesc',
  'removeDoctype',
  'removeEditorsNSData',
  'removeEmptyAttrs',
  'removeEmptyContainers',
  'removeEmptyText',
  'removeHiddenElems',
  'removeMetadata',
  'removeNonInheritableGroupAttrs',
  'removeTitle',
  'removeUnknownsAndDefaults',
  'removeUnusedNS',
  'removeUselessDefs',
  'removeUselessStrokeAndFill',
  'removeViewBox',
  'removeXMLProcInst',
  'sortAttrs',
  'sortDefsChildren',
]

// @keep-sorted
export const PLUGIN_NON_DEFAULT = [
  // required params
  'addAttributesToSVGElement',
  // required params
  'addClassesToSVGElement',
  'cleanupListOfValues',
  'convertOneStopGradients',
  'convertStyleToAttrs',
  'prefixIds',
  // required params
  'removeAttributesBySelector',
  // required params
  'removeAttrs',
  'removeDimensions',
  // required params
  'removeElementsByAttr',
  'removeOffCanvasPaths',
  'removeRasterImages',
  'removeScriptElement',
  'removeStyleElement',
  'removeXlink',
  'removeXMLNS',
  'reusePaths',
]
