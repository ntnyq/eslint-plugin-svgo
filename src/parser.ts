import type { Linter } from 'eslint'

/**
 * @copyright {@link https://github.com/so1ve/eslint-parser-plain}
 */
export const parserPlain: Linter.Parser = {
  meta: {
    name: 'plain-eslint-parser',
  },
  parseForESLint: (code: string) => ({
    ast: {
      body: [],
      comments: [],
      loc: { start: 0, end: code.length },
      range: [0, code.length],
      tokens: [],
      type: 'Program',
    },
    scopeManager: null,
    services: { isPlain: true },
    visitorKeys: {
      Program: [],
    },
  }),
}
