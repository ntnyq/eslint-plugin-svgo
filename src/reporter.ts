/**
 * @copyright {@link https://github.com/antfu/eslint-formatting-reporter}
 */

import { DIFFERENCE, generateDifferences } from 'generate-differences'
import { showInvisibles } from 'show-invisibles'
import type { Rule } from 'eslint'
import type { Difference } from 'generate-differences'

export const messages = {
  [DIFFERENCE.INSERT]: 'Insert `{{ insertText }}`',
  [DIFFERENCE.DELETE]: 'Delete `{{ deleteText }}`',
  [DIFFERENCE.REPLACE]: 'Replace `{{ deleteText }}` with `{{ insertText }}`',
}

function _reportDifference(
  context: Rule.RuleContext,
  difference: Difference,
  rangeOffset = 0,
) {
  const { offset, operation } = difference
  let deleteText = ''
  let insertText = ''

  if ('deleteText' in difference && difference.deleteText) {
    deleteText = difference.deleteText
  }

  if ('insertText' in difference && difference.insertText) {
    insertText = difference.insertText
  }

  const range = [
    offset + rangeOffset,
    offset + rangeOffset + deleteText.length,
  ] as [number, number]
  const [start, end] = range.map(index =>
    context.sourceCode.getLocFromIndex(index),
  )

  context.report({
    messageId: operation,
    data: {
      deleteText: showInvisibles(deleteText),
      insertText: showInvisibles(insertText),
    },
    loc: { start, end },
    fix: fixer => fixer.replaceTextRange(range, insertText),
  })
}

export function reportDifferences(
  context: Rule.RuleContext,
  source: string,
  formatted: string,
  offset = 0,
) {
  if (source === formatted) {
    return
  }

  const differences = generateDifferences(source, formatted)

  for (const difference of differences) {
    _reportDifference(context, difference, offset)
  }
}
