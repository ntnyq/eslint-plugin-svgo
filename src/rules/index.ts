import { svgo } from './svgo'

export const rules = {
  svgo,
}
