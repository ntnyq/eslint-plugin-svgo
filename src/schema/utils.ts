import type { JSONSchema4 } from 'json-schema'

export function createParamsSchema(
  properties: JSONSchema4['properties'] = {},
  additionalProperties = false,
) {
  const schema: JSONSchema4 = {
    type: 'object',
    properties,
    additionalProperties,
  }
  return schema
}

export function createPluginSchema(pluginName: string, params: JSONSchema4) {
  const schema: JSONSchema4 = {
    type: 'object',
    properties: {
      name: {
        type: 'string',
        enum: [pluginName],
      },
      params,
    },
    additionalProperties: false,
  }

  return schema
}
