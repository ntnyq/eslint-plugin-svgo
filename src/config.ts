import { parserPlain } from './parser'
import plugin from '.'
import type { Linter } from 'eslint'
import type { RuleOptions } from '../dts/rule-options'

/**
 * Just options of ESLint config options
 *
 * @see {@link https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects}
 */
export type CreateConfigOptions = Linter.Config<RuleOptions>

/**
 * Shortcut for creating an ESLint config
 *
 * @param options - {@link CreateConfigOptions}
 * @returns An eslint config
 */
export function createConfig(options: CreateConfigOptions = {}) {
  const config: Linter.Config<RuleOptions> = {
    ...options,

    plugins: {
      ...(options.plugins || {}),
      /* v8 ignore start */
      get svgo() {
        return plugin
      },
      /* v8 ignore stop */
    },
    languageOptions: {
      ...(options.languageOptions || {}),
      parser: parserPlain,
    },
  }

  return config
}

export const configs = {
  /**
   * recommended config for eslint-plugin-svgo
   */
  recommended: createConfig({
    name: 'svgo/recommended',
    files: ['**/*.svg'],
    rules: {
      'svgo/svgo': 'error',
    },
  }),
}
