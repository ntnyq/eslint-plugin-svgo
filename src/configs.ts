import { parserPlain } from './parser'
import plugin from '.'
import type { Linter } from 'eslint'
import type { RuleOptions } from '../dts/rule-options'

/**
 * Just options of ESLint config options
 *
 * @see {@link https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects}
 */
export type CreateConfigOptions = Linter.Config<RuleOptions>

/**
 * Shortcut for creating an ESLint config
 *
 * @param options - {@link CreateConfigOptions}
 * @returns An eslint config
 *
 * @deprecated
 */
/* v8 ignore start */
export function createConfig(options: CreateConfigOptions = {}) {
  const config: Linter.Config<RuleOptions> = {
    ...options,

    plugins: {
      ...(options.plugins || {}),
      get svgo() {
        return plugin
      },
    },
    languageOptions: {
      ...(options.languageOptions || {}),
      parser: parserPlain,
    },
  }

  return config
}
/* v8 ignore stop */

/**
 * Recommended config of eslint-plugin-svgo
 */
export const recommended: Linter.Config = {
  name: 'svgo/recommended',
  files: ['**/*.svg'],
  plugins: {
    /* v8 ignore start */
    get svgo() {
      return plugin
    },
    /* v8 ignore stop */
  },
  languageOptions: {
    parser: parserPlain,
  },
  rules: {
    'svgo/svgo': 'error',
  },
}

export const configs = {
  recommended,
}
