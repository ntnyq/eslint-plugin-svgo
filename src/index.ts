import { config } from './config'
import { meta } from './meta'
import { rules } from './rules'
import type { ESLint } from 'eslint'

export const plugin = {
  meta,
  rules,
  configs: {
    recommended: config(),
  },
} satisfies ESLint.Plugin

export * from './meta'
export * from './rules'
export * from './config'
export * from './parser'

export default plugin
