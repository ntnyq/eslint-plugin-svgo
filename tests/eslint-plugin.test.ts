import { ESLint } from 'eslint'
import { glob } from 'tinyglobby'
import { expect, it } from 'vitest'
import { resolve } from '../scripts/utils'
import { plugin as pluginSvgo } from '../src'

const TEST_CWD = resolve('tests/fixtures/eslint-plugin')

it('should lint work', async () => {
  const files = await glob('**/*.svg', { cwd: TEST_CWD, onlyFiles: true })
  const eslint = new ESLint({
    overrideConfigFile: true,
    overrideConfig: [pluginSvgo.configs.recommended],
    cwd: TEST_CWD,
    ignore: false,
  })

  const results: ESLint.LintResult[] = await eslint.lintFiles(files)

  expect(results.length).toBe(files.length)

  results.forEach((result, idx) => {
    expect(result.messages).toMatchSnapshot(files[idx])
  })
})
