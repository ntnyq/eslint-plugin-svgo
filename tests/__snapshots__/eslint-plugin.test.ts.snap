// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should lint work > cleanup-ids.svg 1`] = `
[
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 12,
    "fix": {
      "range": [
        0,
        374,
      ],
      "text": "<svg xmlns="http://www.w3.org/2000/svg" width="150" viewBox="0 0 150 100"><defs><circle id="a" cx="50" cy="50" r="50" fill="#000"/></defs><use href="#a" transform="skewX(16)"/></svg>",
    },
    "line": 1,
    "message": "Replace \`<?xml·version="1.0"·encoding="UTF-8"?>⏎<svg·xmlns="http://www.w3.org/2000/svg"·viewBox="·0·0··150·100·"·width="150">⏎··<!--·Created·with·love!·-->⏎··<defs>⏎····<ellipse·cx="50"·cy="50.0"·rx="50.00"·ry="auto"·fill="black"·id="circle"/>⏎··</defs>⏎··<g>⏎····<use·href="#circle"·transform="skewX(16)"/>⏎····<rect·id="useless"·width="0"·height="0"·fill="#ff0000"/>⏎··</g>⏎</svg>⏎\` with \`<svg·xmlns="http://www.w3.org/2000/svg"·width="150"·viewBox="0·0·150·100"><defs><circle·id="a"·cx="50"·cy="50"·r="50"·fill="#000"/></defs><use·href="#a"·transform="skewX(16)"/></svg>\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "svgo/svgo",
    "severity": 2,
  },
]
`;

exports[`should lint work > invalid.svg 1`] = `
[
  {
    "column": 1,
    "endColumn": 1,
    "endLine": 2,
    "line": 2,
    "message": "SvgoParserError: Unclosed root tag",
    "nodeType": null,
    "ruleId": "svgo/svgo",
    "severity": 2,
  },
]
`;
